test/distributed/test_c10d.py

```diff
@@ -4019,6 +4019,54 @@ def allreduce_hook_gloo(
 
         self._run_and_verify_sparse_gradients(vanilla_model, ddp_model)
 
+    class AcceptsParam(torch.nn.Module):
+        def __init__(self, p, factor):
+            super().__init__()
+            self.a = p
+            self.f = factor
+
+        def forward(self, input):
+            return input + self.a * self.f
+
+    @requires_nccl()
+    def test_ddp_weight_sharing(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        size = 2048 * 2048
+        dev = self.rank
+        world = self.world_size
+
+        p = torch.nn.Parameter(torch.randn(size, requires_grad=True))
+
+        for try_set_to_none, use_bucket_view in product((False, True), (False, True)):
+            m = torch.nn.Sequential(self.AcceptsParam(p, dev + 1),
+                                    self.AcceptsParam(p, dev + 1)).cuda(dev)
+
+            m = torch.nn.parallel.DistributedDataParallel(m,
+                                                          bucket_cap_mb=1,
+                                                          gradient_as_bucket_view=use_bucket_view,
+                                                          device_ids=[dev],
+                                                          process_group=process_group)
+
+            for i in range(3):
+                m.zero_grad(set_to_none=try_set_to_none)
+                m(1).sum().backward()
+
+                analytic = torch.full_like(p, 2. * (world * (world + 1.) / 2.) / world, device=dev)
```

Comment:
just for my understanding: 2. * (world * (world + 1.) / 2.) / world = world +1, right? also, would you please help commenting why the analytic is calculated in this way?

Reply:
Do you mean, where the math comes from, or why I didn't simplify it?  I didn't simplify it because I forgot :P

Reply:
Do you want me to add a comment how I got the analytic value based on what each layer's backward does?

Reply:
yes, I'm a little confused about the formula, also, feel like it can be simplified as well?

Reply:
Added explanation. If you don't mind i'd rather keep it unsimplified, it's easier to map to the explanation that way.

Reply:
sounds good!